Add an input_datetime

tinloaf opened the pull request:
## Description:

This is a Work-in-Progress proposal. It's not ready for merging yet, but I'd appreciate some feedback.

This adds an input_datetime component allowing to set a date, a time, or both via the frontend. This date / time can then be used by automations etc. to trigger certain events at certain times, e.g. facilitating alarm clocks, time-setpoints for the opening / closing of covers, etc. pp.

Some people have requested this in the forum [0], and there are some hacks around mimicking a 'set time' functionality via a combination of input sliders or input selects (e.g., [1], [2]). Thus, I think there is a demand for such an input (I'd like to use it myself for an alarm clock).

There are some design decision that I'm not sure about. The most important one being: Depending on whether the input is configured to have time, date or both, the status of the input is a date, time or datetime object. I'm not sure whether to default to datetime in all cases. The advantage would be that the class of the state object does not depend on the configuration, the disadvantage would be that I'd need some dummy values for date / time when it's not used.

This is the accompanying polymer frontend PR: https://github.com/home-assistant/home-assistant-polymer/pull/418 

[0] https://community.home-assistant.io/t/date-time-picker/9062
[1] https://community.home-assistant.io/t/creating-a-alarm-clock/410
[2] https://community.home-assistant.io/t/alarm-clock-with-sonos-philips-hue/23421

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#3390

## Example entry for `configuration.yaml` (if applicable):
```yaml
input_datetime:
  test:
    name: 'Test'
    has_time: true
    has_date: true
  test_date:
    name: 'Test Date'
    has_time: false
    has_date: true
  test_time:
    name: 'Test Time'
    has_time: true
    has_date: false
```

## Checklist:

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code does not interact with devices:
  - [x] Local tests with `tox` run successfully. **Your PR cannot be merged unless tests pass**
  - [x] Tests have been added to verify that the new code works.

tinloaf committed:
Initial proposal for the input_datetime

tinloaf commented:
Whoop, sorry, should really have linted before creating the PR.

tinloaf committed:
Linting

tinloaf commented:
Here's some screenshots, of the 'More Info' dialog including the two pickers for date and time (which definitely needs some styling done…), of the date picker, and of a card containing a datetime, date and time input:
![card](https://user-images.githubusercontent.com/442687/30077518-63ebea50-927c-11e7-8e55-648a83a2838b.png)
![more_options](https://user-images.githubusercontent.com/442687/30077516-63eb9b4a-927c-11e7-8c44-353c9299ac12.png)
![picker](https://user-images.githubusercontent.com/442687/30077517-63ebc886-927c-11e7-9db5-e33c4e97eded.png)

dale3h commented:
At this point, before moving further with other `input_xxxxxxx` components, would it be a good idea to break this out into a single component (`input`) with multiple platforms (`boolean`, `select`, `text`, `number`, `datetime`)?

tinloaf commented:
@dale3h I actually also thought about that. However, the different inputs behave very differently, I think. For example have a look at the services provided by input select (select next / previous…) and the input boolean (turn on / turn off). While for the input boolean, this could probably be covered by a generic 'set_value', turn on / off is more in accordance with light / switches, I think.

Also, as far as I see it, the UI elements are determined per domain, i.e., would be the same for all inputs. I can hardly imagine some UI that's fitting for a 'select' input as well as for a date / time input or a text input.

As far as I see this, a 'component' describes a certain way of behavior - basically, an interface to the user, while a platform describes the technical way to achieve this. With the inputs, it's somewhat the other way round: The technical way of achieving things is kind of similar (store some state, make it editable), but the interaction with the user is very different.

tinloaf committed:
Further linting, don't define time validation twice

tinloaf committed:
Make pylint *and* flake8 happy at the same time

tinloaf committed:
Move todos to the PR to make lint happy

tinloaf committed:
Actually validate the type of date/time

tinloaf committed:
First testing

tinloaf committed:
Linting

pvizeli commented:
Yeah the UI work with component part to deside the correct card. I think a component/platform solution is hard for the frontent. Maybe we can convert it later if the UI is ready for that.

pvizeli commented on homeassistant/components/input_datetime.py:
```diff
@@ -0,0 +1,214 @@
+
+    if not has_date and not has_time:
+        raise vol.Invalid("Input Datetime must have at least date or time!")
+    return cfg
```
use vol.Exclusive (I'm not 100% sure) but the already exists a validate options to handle this case

tinloaf replied:
According to https://github.com/alecthomas/voluptuous/issues/126 , this functionality does not currently exist. vol.Exclusive() is 'XOR', we need 'OR', basically. The solution suggested in the issue is to have two schemata: one for validating the keys, one for validating the values. I could probably adapt this to something like this (with different names / constants): 
```
has_date_or_time_schema = Schema({
    Required(
        Any('has_time', 'has_date'),
        msg="Must have at least one of time or date!"): True
}, extra=ALLOW_EXTRA)

config_values_schema = Schema({
    Optional('has_time'): bool,
    Optional('has_date'): bool,
    … (all the rest) …
})

config_schema = All(has_date_or_time_schema, config_values_schema)
```
Would you consider that preferable? Or did I miss some better way?

MartinHjelmare replied:
Maybe use https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/config_validation.py#L45?

tinloaf replied:
Ah, now that I try to use `has_at_least_one_key`: What I actually need to check is not that at least one of the two keys is present, but that at least one of the two is present *and* set to True. Otherwise, this allows configuring an input_datetime that carries neither time nor date.

I think the above example achieves that, the current code in the PR also does, but with `has_at_least_one_key` that is not possible. A third possibility that I see (additional to what's currently there and what I outlined above):

- Add an `as_at_least_one_key_value` to config_validation.py which takes a list of (key, value) pairs and validates that at least one of those is present in the dict.
- Use that with `[(CONF_HAS_TIME, True), (CONF_HAS_DATE, True)]` in the schema

Would that be an acceptable solution?

pvizeli commented on homeassistant/components/input_datetime.py:
```diff
@@ -0,0 +1,214 @@
+            vol.Required(CONF_HAS_TIME): cv.boolean,
+            vol.Optional(CONF_ICON): cv.icon,
+        }, _cv_input_datetime)})
+}, required=True, extra=vol.ALLOW_EXTRA)
```
remove `required`

pvizeli commented on homeassistant/components/input_datetime.py:
```diff
@@ -0,0 +1,214 @@
+        else:
+            restore_val = dt_util.now()
+
+        print("Type is: " + str(type(restore_val)))
```
remove this

pvizeli commented on homeassistant/components/input_datetime.py:
```diff
@@ -0,0 +1,214 @@
+        return attrs
+
+    @asyncio.coroutine
+    def async_set_datetime(self, date_val, time_val):
```
make no coroutine only a @callback

pvizeli commented on homeassistant/components/input_datetime.py:
```diff
@@ -0,0 +1,214 @@
+            self._current_datetime = datetime.datetime.combine(date_val,
+                                                               time_val)
+
+        yield from self.async_update_ha_state()
```
move that into handler function

tinloaf replied:
I don't understand: isn't `async_set_datetime` already the handler function? It directly sets the internal state (`self._current_datetime`), no further handler is called, or am I missing something? Or should I add another handler function, schedule a call to this handler function at the end of `async_set_datetime`, and then put `self.async_update_ha_state()` into it? 

I need to call `self.async_update_ha_state()` every time I modify the internal state of an entity, right? Such that the state machine and the frontend update their state?

tinloaf committed:
Address code review issues

pvizeli commented on homeassistant/components/input_datetime.py:
```diff
@@ -0,0 +1,214 @@
+
+        tasks = []
+        for input_datetime in target_inputs:
+            print("Setting dt for " + str(input_datetime))
```
Remove this

pvizeli commented on homeassistant/components/input_datetime.py:
```diff
@@ -0,0 +1,214 @@
+        tasks = []
+        for input_datetime in target_inputs:
+            print("Setting dt for " + str(input_datetime))
+            tasks.append(
```
call async_set_datetime() and add/append async_update_ha_state() as task

pvizeli commented on homeassistant/components/input_datetime.py:
```diff
@@ -0,0 +1,214 @@
+            self._current_datetime = datetime.datetime.combine(date_val,
+                                                               time_val)
+
+        yield from self.async_update_ha_state()
```
Remove this

pvizeli reviewed (changes requested):
make the `async_set_datetime` as coroutine. I update the logic later for that input components.

pvizeli commented on homeassistant/components/input_datetime.py:
```diff
@@ -0,0 +1,214 @@
+        return attrs
+
+    @callback
+    def async_set_datetime(self, date_val, time_val):
```
Don't validate stuff inside this function. That will be done with vololuptus

tinloaf replied:
The problem is that inside the service schema, I need to make date and time both optional, because there can be `input_datetime`s that dont carry date / time. But, depending on the actual configuration of the entity, they need time and / or date. I don't know that at the time I register the service and therefore the service schema.

Thus, I cannot assume that either `date_val` or `time_val` is not `None` in this method.

Another option I see would be to break the service into three: SET_DATETIME, SET_DATE and SET_TIME, which then have required 'date' and/or 'time' data associated with them. Is that a better solution? It would mean that callers such as the frontend need a bit more logic to figure out which service to call…

balloob replied:
Wouldn't you be able to use the `has_at_least_one_key` validation then? https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/config_validation.py#L45

tinloaf replied:
@balloob You mean in the schema for `SET_DATETIME` without breaking it into three? The problem is that the question which of `CONF_HAS_TIME` and/or `CONF_HAS_DATE` are set to `True` determines exactly which keys must be there. For example: Say you configure the input like this:

```
… 
  has_time: True
  has_date: False
…
```

and you call the service with `{'date': '2017-01-01'}`. Now the service schema with `has_at_least_one_key(['date', 'time'])` happily validates, but it's the service can't work with this data.

balloob replied:
In that case you can use `vol.Any` ([example](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/config_validation.py#L190-L191))

tinloaf committed:
Code review: Remove forgotten print()s

tinloaf committed:
Make set_datetime a coroutine

tinloaf committed:
Create contains_at_least_one_key_value CV method, use it

tinloaf commented:
Regarding the validation that at least one of `CONF_HAS_DATE` and `CONF_HAS_TIME` must be set to `True`: I just created a `has_at_least_one_key_value` method in config_validation.py (which basically works the same as `has_at_least_one_key`) and use it to validate instead of having a helper method in input_datetime.py. Does that look like a feasible approach?

tinloaf committed:
Add timestamp to the attributes

tinloaf committed:
Test and fix corner case where restore data is bogus

pvizeli committed:
Add FIXME

pvizeli commented on homeassistant/helpers/config_validation.py:
```diff
@@ -144,6 +160,53 @@ def icon(value):
+    return time_val
+
+
+def time_time(time_val: time_sys) -> time_sys:
```
merge `time_str` and `time_time` into `time` and add unittests

pvizeli commented on homeassistant/helpers/config_validation.py:
```diff
@@ -144,6 +160,53 @@ def icon(value):
+    return date_val
+
+
+def date_date(date_val: date_sys) -> date_sys:
```
merge `date_str` and `date_date` into `date` and add unittests

pvizeli reviewed (changes requested):
I add some FIXME. Solve that and we can merge it 👍

tinloaf committed:
Fix date/time setting

tinloaf committed:
Fix Validation

tinloaf commented:
I've fixed the places with FIXMEs (I hope the way I validate the service data is what you had in mind?) and also merged the time / date validators and added tests.

tinloaf committed:
Merge date / time validation, add tests

tinloaf commented:
I just created the documentation PR. I think this is ready for merging from my side now.

tinloaf commented:
@pvizeli Did I address your change requests? Or is something left?

pvizeli commented on homeassistant/components/input_datetime.py:
```diff
@@ -0,0 +1,237 @@
+            vol.Required(CONF_HAS_TIME): cv.boolean,
+            vol.Optional(CONF_ICON): cv.icon,
+            vol.Optional(CONF_INITIAL,
+                         default=datetime.datetime(1970, 1, 1)): cv.datetime,
```
That should be not have a default value. If we have a initial value we reset every boot this to initial state like other input types.

pvizeli commented on homeassistant/components/input_datetime.py:
```diff
@@ -0,0 +1,237 @@
+            except vol.Invalid as ex:
+                _LOGGER.error("Invalid service data for "
+                              "input_datetime.set_datetime: %s",
+                              humanize_error(call.data, ex))
```
Remove this. Make it simple like:

```
time = call.data.get(ATTR_DATE)
date = call.data.get(ATTR_TIME)
if (x.has_date() and not time) or ():
   _LOGGER.error()
   continue
tasks.append(x.async_set_datetime(date, time)
```

pvizeli reviewed (changes requested):
Looks good

tinloaf committed:
Simplify service data validation

tinloaf committed:
No default for initial state, allow 'unknown' as state

pvizeli committed:
cleanup

pvizeli committed:
fix schema

bfranske commented:
There are a couple of semi-related things which could probably still use some cleanup on this. The first is that it's a little awkward to fire a service call for both a change in the hour and the minute when setting a time. This potentially fires automations at unexpected times because an automation could match a time after you have changed the hour but before changing the minute.

The second thing is that although there is a calendar date picker there is not really a usable time picker, especially on mobile devices. Instead of time entry by text field it would be nice to have a clock face picker like https://weareoutman.github.io/clockpicker/ This would be especially useful for mobile devices but is also pretty convenient on PCs.

MartinHjelmare commented:
Please open an issue if you suspect a bug.

If you want to suggest an enhancement please open a feature request in [the Feature Requests section](https://community.home-assistant.io/c/feature-requests) of our community forum.

Merged PRs should not be used for enhancement discussion or bug reports. If you've found a bug it's ok to make a review with inline comments and link to an issue that reports the bug.

Thanks!
